switchbot/__init__.py

```diff
@@ -93,119 +149,156 @@ def _writekey(self, key) -> bool:
         _LOGGER.debug("Sending command to switchbot %s", command)
         try:
-            self._connect()
-            send_success = self._writekey(command)
+            with CONNECT_LOCK:
+                self._connect()
+                send_success = self._writekey(command)
         except bluepy.btle.BTLEException:
-            _LOGGER.warning("Error talking to Switchbot.", exc_info=True)
+            _LOGGER.warning("Error talking to Switchbot", exc_info=True)
         finally:
             self._disconnect()
         if send_success:
             return True
         if retry < 1:
             _LOGGER.error(
-                "Switchbot communication failed. Stopping trying.", exc_info=True
+                "Switchbot communication failed. Stopping trying", exc_info=True
             )
             return False
-        _LOGGER.warning(
-            "Cannot connect to Switchbot. Retrying (remaining: %d)...", retry
-        )
+        _LOGGER.warning("Cannot connect to Switchbot. Retrying (remaining: %d)", retry)
         time.sleep(DEFAULT_RETRY_TIMEOUT)
         return self._sendcommand(key, retry - 1)
 
-    def get_servicedata(self, retry=DEFAULT_RETRY_COUNT, scan_timeout=5) -> bytearray:
-        """Get BTLE 16b Service Data,
-        returns after the given timeout period in seconds."""
+    def discover(self, retry=DEFAULT_RETRY_COUNT, scan_timeout=5) -> dict:
+        """Find switchbot devices and their advertisement data."""
+
         devices = None
 
-        waiting_time = self._time_between_update_command - time.time()
-        if waiting_time > 0:
-            time.sleep(waiting_time)
         try:
             devices = bluepy.btle.Scanner().scan(scan_timeout)
 
         except bluepy.btle.BTLEManagementError:
-            _LOGGER.warning("Error updating Switchbot.", exc_info=True)
+            _LOGGER.error("Error scanning for switchbot devices", exc_info=True)
 
         if devices is None:
             if retry < 1:
                 _LOGGER.error(
-                    "Switchbot update failed. Stopping trying.", exc_info=True
+                    "Scanning for Switchbot devices failed. Stop trying", exc_info=True
                 )
                 return None
 
             _LOGGER.warning(
-                "Cannot update Switchbot. Retrying (remaining: %d)...", retry
+                "Error scanning for Switchbot devices. Retrying (remaining: %d)",
+                retry,
             )
             time.sleep(DEFAULT_RETRY_TIMEOUT)
-            return self.get_servicedata(retry - 1, scan_timeout)
-
-        for device in devices:
-            if self._mac.lower() == device.addr.lower():
-                for (adtype, _, value) in device.getScanData():
+            return self.discover(retry - 1, scan_timeout)
+
+        for dev in devices:
+            if dev.getValueText(7) == UUID:
+                dev_id = dev.addr.replace(":", "")
+                self._all_services_data[dev_id] = {}
```

Comment:
The SwitchbotDevice class should only keep information about one device

Reply:
I could split it up in different classes? Considering a singe scan returns all data I thought this is the most efficient way of doing it without breaking any 3rd party integrations that makes use of this api. 

Bluetooth adapters can't do many operations at a time and need to be used sparingly. (especially on low power/low cost devices like raspberry pi)

Reply:
Great to reduce the number of scans.
I do not think it then make sense to have the discover function as part of the SwitchbotDevice

Reply:
This release will anyway have breaking changes

Reply:
I have moved discovery to a new Class.